Compression.md

```diff
@@ -0,0 +1,145 @@
+
+This document assumes the reader is aware of the fundamentals of the [WebSocket protocol](https://tools.ietf.org/html/rfc6455).
+
+### Table of contents
+1. [WebSocket extensions](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#1-websocket-extensions)
+2. [WebSocket compression - the `permessage-deflate` algorithm](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#2-the-permessage-deflate-algorithm)
+3. [An implementation of permessage-deflate based on swift-nio](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#3-an-implementation-of-permessage-deflate-based-on-swift-nio)
+4. [Developer notes](https://gist.github.com/pushkarnk/aea1be88c7c7283fcfe4615df35ed857#4-developer-notes)
+
+### 1. WebSocket extensions
+The WebSocket protocol has a provision for servers to configure protocol extensions, and for clients to request these extensions from the servers. A client notifies about extensions it is interested in through a `negotiation offer` using the `Sec-WebSocket-Extension` header. A server may or may not support the extensions requested by the client. Through a `negotiation response`, the server notifies the client of the extensions that the server agrees upon. The negotiation response offer and response may also include extension-specific parameters. Once an extension is agreed upon, the client and server must invoke the extension from their respective WebSocket implementations.
+
+WebSocket compression is a WebSocket extension.
+
+### 2. WebSocket Compression: the permessage-deflate algorithm
+Permessage-deflate is a WebSocket extension defined by [RFC7692](https://tools.ietf.org/html/rfc7692) which provides a specification for the compression functionality. It defines the negotiation process and a compression algorithm called `DEFLATE`. 
+
+A `permessage-deflate` negotiation happens during the upgrade request from HTTP to WebSocket. A `permessage-deflate` negotiation offer has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters. There are four extension parameters defined for WebSocket compression:
+ - `client_no_context_takeover`, `server_no_context_takeover`
+ - `client_max_window_bits`, `server_max_window_bits`
+We shall revisit these parameters in the future sections, where we discuss their use and effects.
+
+A permessage-deflate negotiation response has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters, agreed upon by the server. The headers in the negotiation response are the final word on how the data compression/decompression will be done between the client and server. Data compressed by the client must be decompressed by the server and vice versa. The client and server must adopt the same compression/decompression configuration parameters. We shall deal with this in detail in the future sections.
```

Comment:
We shall deal with this in detail in the future sections. -> "We will talk about this in detail later in this document."